src/tools/perf/perftest_mad.c

```diff
@@ -0,0 +1,694 @@
+
+    lid = rte_group->dst_port.lid;
+    ucs_debug("MAD: Accept: remote lid:%d/0x%02x", lid, lid);
+
+    /* Import received message size list */
+    size = sizeof(*ctx->params.super.msg_size_list) *
+           ctx->params.super.msg_size_cnt;
+    ptr  = calloc(1, size);
+    if (ptr == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+    memcpy(&ctx->params, buf, sizeof(ctx->params));
+    ctx->params.super.msg_size_list = ptr;
+
+    memcpy(ctx->params.super.msg_size_list, buf + sizeof(ctx->params), size);
+
+    return perftest_mad_send(rte_group, &mad_magic, sizeof(mad_magic));
+}
+
+static ucs_status_t perftest_mad_connect(perftest_mad_rte_group_t *rte_group,
+                                         struct perftest_context *ctx)
+{
+    int iovcnt = 1;
+    ucs_status_t status;
+    size_t size;
+    struct iovec iov[2];
+
+    iov[0].iov_base = &ctx->params;
+    iov[0].iov_len  = sizeof(ctx->params);
+
+    size  = sizeof(*ctx->params.super.msg_size_list);
+    size *= ctx->params.super.msg_size_cnt;
+
+    if (size != 0) {
+        iov[1].iov_base = ctx->params.super.msg_size_list;
+        iov[1].iov_len  = size;
+        iovcnt++;
+    }
+    status = perftest_mad_sendv(rte_group, iov, iovcnt);
+    if (status != UCS_OK) {
+        return status;
+    }
+    return perftest_mad_recv_magic(rte_group, mad_magic);
+}
+
+static void perftest_mad_set_logging(void)
+{
+#define IB_DEBUG_LEVEL 10
+    if (ucs_log_is_enabled(UCS_LOG_LEVEL_DEBUG)) {
+        ibdebug = IB_DEBUG_LEVEL; /* extern variable from mad headers */
+        umad_debug(IB_DEBUG_LEVEL);
+    }
+}
+
+ucs_status_t setup_mad_rte(struct perftest_context *ctx)
+{
+    int is_server = !ctx->server_addr;
+    perftest_mad_rte_group_t *rte_group;
+    ucs_status_t ret;
+
+    rte_group = calloc(1, sizeof(*rte_group));
+    if (rte_group == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    perftest_mad_set_logging();
+
+    rte_group->is_server = is_server;
+    rte_group->mad_port  = perftest_mad_open(ctx->ib.ca, ctx->ib.ca_port,
+                                             is_server);
+    if (rte_group->mad_port == 0) {
+        ucs_error("MAD: %s: cannot open port '%s:%d'",
+                  is_server ? "server" : "client", ctx->ib.ca, ctx->ib.ca_port);
+        goto fail;
+    }
+
+    if (is_server) {
+        ret = perftest_mad_accept(rte_group, ctx);
+    } else {
+        /* Lookup server if needed */
+        ret = perftest_mad_get_portid(ctx->ib.ca, ctx->ib.ca_port,
+                                      ctx->server_addr, rte_group->mad_port,
+                                      &rte_group->dst_port);
+        if (ret != UCS_OK) {
+            ucs_error("MAD: client: cannot get port as: '%s:%d' -> '%s'",
+                      ctx->ib.ca, ctx->ib.ca_port, ctx->server_addr);
+            goto fail;
+        }
+
+        /* Try to connect to it */
+        ret = perftest_mad_connect(rte_group, ctx);
+    }
+
+    if (ret != UCS_OK) {
```

Comment:
IMO better do it after each function call

Reply:
removed and added if(){} after perftest_mad_connect and after perftest_mad_accept